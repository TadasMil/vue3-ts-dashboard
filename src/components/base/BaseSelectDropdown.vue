<script setup lang="ts" name="BaseSelectDropdown" generic="T extends IOption">
import ArrowDown from 'virtual:icons/mdi/chevron-down'

import { IOption } from '@/types/selectDropdown'

const { options } = defineProps<{
  options: T[]
  modelValue: T | null
}>()

const inputValue = defineModel<T | null>('modelValue', { default: null })
</script>

<template>
  <div class="base-dropdown">
    <select v-model="inputValue" class="base-dropdown__select" data-testid="select-dropdown">
      <option :value="null" selected disabled hidden>Sort by</option>

      <option
        v-for="option in options"
        :key="option.id"
        :value="option"
        data-testid="select-dropdown-option"
      >
        {{ option.name }}
      </option>
    </select>

    <ArrowDown class="base-dropdown__arrow" />
  </div>
</template>

<style scoped>
.base-dropdown {
  display: flex;
  width: 145px;
  justify-content: space-between;
  align-items: center;
  flex-shrink: 0;
  border-radius: 4px;
  border: 1px solid #efefef;
  background: #fff;
  position: relative;
}

.base-dropdown__select {
  width: 100%;
  border: none;
  padding: 6px;
  background: transparent;
  padding-right: 24px;
  color: var(--color-secondary);

  appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
}

.base-dropdown__select:focus {
  outline: none;
  box-shadow: 0 0 0 2px rgba(21, 156, 228, 0.4);
  border-radius: 4px;
}

.base-dropdown__arrow {
  position: absolute;
  right: 6px;
  top: 50%;
  transform: translateY(-50%);
  pointer-events: none;
  color: var(--color-highlight);
}
</style>
