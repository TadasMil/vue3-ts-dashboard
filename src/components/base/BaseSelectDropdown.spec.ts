import { fireEvent, render } from '@testing-library/vue'
import { describe, expect, it } from 'vitest'

import '@testing-library/jest-dom'
import BaseSelectDropdown from './BaseSelectDropdown.vue'

describe('BaseSelectDropdown.vue', () => {
  const mockOptions = ['Option 1', 'Option 2']

  it('renders the default "Sort by" option', async () => {
    const { html } = render(BaseSelectDropdown, {
      props: {
        options: mockOptions
      }
    })

    expect(html()).toMatchSnapshot()
  })

  it('renders provided options and updates the model value when an option is selected', async () => {
    const { getByTestId } = render(BaseSelectDropdown, {
      props: {
        options: mockOptions
      }
    })

    const select = getByTestId('select-dropdown') as HTMLSelectElement

    const optionToSelect = mockOptions[0]

    await fireEvent.click(select, { target: { value: optionToSelect } })

    expect(select.value).toBe(optionToSelect)
  })
})
