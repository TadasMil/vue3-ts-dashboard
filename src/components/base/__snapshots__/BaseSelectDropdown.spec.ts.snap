// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseSelectDropdown.vue > renders the default "Sort by" option 1`] = `
"<div data-v-7c7446ef=\\"\\" class=\\"base-dropdown\\"><select data-v-7c7446ef=\\"\\" class=\\"base-dropdown__select\\" data-testid=\\"select-dropdown\\">
    <option data-v-7c7446ef=\\"\\" disabled=\\"\\" hidden=\\"\\">Sort by</option>
    <option data-v-7c7446ef=\\"\\" data-testid=\\"select-dropdown-option\\" value=\\"Option 1\\"></option>
    <option data-v-7c7446ef=\\"\\" data-testid=\\"select-dropdown-option\\" value=\\"Option 2\\"></option>
  </select><svg data-v-7c7446ef=\\"\\" viewBox=\\"0 0 24 24\\" width=\\"1.2em\\" height=\\"1.2em\\" class=\\"base-dropdown__arrow\\">
    <path fill=\\"currentColor\\" d=\\"M7.41 8.58L12 13.17l4.59-4.59L18 10l-6 6l-6-6l1.41-1.42Z\\"></path>
  </svg></div>"
`;
